src/adapter/src/frontend_peek.rs

```diff
@@ -0,0 +1,613 @@
+            Some(
+                _determination @ TimestampDetermination {
+                    timestamp_context: TimestampContext::TimelineTimestamp { .. },
+                    ..
+                },
+            ) if in_immediate_multi_stmt_txn => {
+                // TODO(peek-seq): handle multi-statement transactions: return the above
+                // `determination` and the holds from txn_read_holds.
+                return Ok(None);
+            }
+            _ => {
+                let determine_bundle = if in_immediate_multi_stmt_txn {
+                    // TODO(peek-seq): handle multi-statement transactions
+                    // needs timedomain_for, which needs DataflowBuilder / index oracle / sufficient_collections
+                    debug!(
+                        "Bailing out from try_frontend_peek_inner, because of in_immediate_multi_stmt_txn"
+                    );
+                    return Ok(None);
+                } else {
+                    // If not in a transaction, use the source.
+                    &input_id_bundle
+                };
+                let (determination, read_holds) = self
+                    .frontend_determine_timestamp(
+                        catalog.state(),
+                        session,
+                        determine_bundle,
+                        &select_plan.when,
+                        target_cluster_id,
+                        &timeline_context,
+                        oracle_read_ts,
+                        real_time_recency_ts,
+                    )
+                    .await?;
+                (determination, read_holds)
+            }
+        };
+
+        // The old peek sequencing's sequence_peek_timestamp does two more things here:
+        // the txn_read_holds stuff and session.add_transaction_ops. We do these later in the new
+        // peek sequencing code, because at this point we might still bail out from the new peek
+        // sequencing, in which case we don't want the mentioned side effects to happen.
+
+        // # From peek_optimize
+
+        // Generate data structures that can be moved to another task where we will perform possibly
+        // expensive optimizations.
+        let timestamp_context = determination.timestamp_context.clone();
+        if session.vars().enable_session_cardinality_estimates() {
+            debug!(
+                "Bailing out from try_frontend_peek_inner, because of enable_session_cardinality_estimates"
+            );
+            return Ok(None);
+        }
+        let stats = Box::new(EmptyStatisticsOracle);
+        let session_meta = session.meta();
+        let now = catalog.config().now.clone();
+        let select_plan = select_plan.clone();
+
+        // TODO(peek-seq): if explain_ctx.needs_plan_insights() ...
+
+        let span = Span::current();
+
+        let (global_lir_plan, _optimization_finished_at) = match mz_ore::task::spawn_blocking(
```

Comment:
Why are we spawning a new task here?

Reply:
There is a tricky reason for this: the important thing here is the `_blocking` part: this puts the task on a special kind of tokio thread (a so-called blocking thread), where it's ok to run long-running-but-not-yielding code (the optimizer). Running such code in a normal task would block other tasks from running when all the normal worker threads (so-called core threads) are filled up. In contrast, for a blocking task, tokio spins up more blocking threads as needed. See more explanation here: https://docs.rs/tokio/latest/tokio/index.html#cpu-bound-tasks-and-blocking-code